### ISSUE TITLE
unable to bind object to blazer <InputSelect>

### ISSUE THREAD

**Issue by gokulgopalakrishnan** [dotnet/aspnetcore#26685]
### Describe the bug

Trying to bind object of a selected value to a parameter property from select dropdown but getting an exception 

`Unhandled exception rendering component: Microsoft.AspNetCore.Components.Forms.InputSelect`1[MyApp.Models.Country] does not support the type 'MyApp.Models.Country'.` 

Refer the snippet below. I'm trying to bind the selected country object (which has `Id` and `Name`) to the country object of address parameter. 

This is possible in angular though. Is this something will be supported in future releases or I'm trying to code wrong way, any help would be appreciated, thanks!

### To Reproduce

```
<InputSelect @bind-Value="@Address.Country" class="form-control">
                @if (Countries != null)
                {
                    @foreach (var country in countries)
                    {
                        <option value="@country">@country.Name</option>
                    }
                }               
</InputSelect>

@code{
    [Parameter]
    public Address Address
    {
        get;
        set;
    }

    [Parameter]
    public List<Countries> Countries
    {
        get;
        set;
    }
}
```

### Exceptions (if any)
`Unhandled exception rendering component: Microsoft.AspNetCore.Components.Forms.InputSelect`1[MyApp.Models.Country] does not support the type 'MyApp.Models.Country'.` 

### Further technical details

I'm using latest version of VSCode
```
dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.1.402
 Commit:    9b5de826fd

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  10.15
 OS Platform: Darwin
 RID:         osx.10.15-x64
 Base Path:   /usr/local/share/dotnet/sdk/3.1.402/

Host (useful for support):
  Version: 3.1.8
  Commit:  9c1330dedd

.NET Core SDKs installed:
  3.1.102 [/usr/local/share/dotnet/sdk]
  3.1.402 [/usr/local/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.NETCore.App 3.1.2 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.8 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
```

**Comment by SteveSandersonMS** [comment]
`InputSelect` is a wrapper around the HTML `<select>` element, which in turn wants to match up its current value with the `value` attribute from an `<option>`. As such there's no native concept of the value being a complex-typed object, and Blazor doesn't try to layer on any such concept.

So you'll need the value you're binding to be a primitive type (`string`, `int`, GUID, etc.). Normally you'd want the value to correspond to some kind of unique identifier / primary key on the entity, presumably the `Id` property of your `Country` type:

```razor
    <InputSelect @bind-Value="@SelectedCountryId" class="form-control">
        @if (Countries != null)
        {
            @foreach (var country in Countries)
            {
                <option value="@country.Id">@country.Name</option>
            }
        }
    </InputSelect>
```

... and then in your `@code` block, define `SelectedCountryId` like this:

```cs
    private int SelectedCountryId
    {
        get => Address.Country.Id;
        set => Address.Country = Countries.Single(c => c.Id == value);
    }
```

**Comment by gokulgopalakrishnan** [comment]
thank you @SteveSandersonMS , I did try changing the binding to primitive types before, it worked.  I was just curious and wanted to try this way. you see why I did this because in the backend the `Address` entity has a complex object `Country` with `Id` and `Name` as a foreign key so I thought it would be easier to bind the same in the Blazor upon select from dropdown and pass it as input to the API. Anyways I made it work by using the selected `Id` as lookup to `GetCountries` API and bind it to `Address`.`Country`.
